bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/OpenhabGraalJSScriptEngine.java

```diff
@@ -0,0 +1,132 @@
+import org.slf4j.LoggerFactory;
+
+import com.oracle.truffle.js.scriptengine.GraalJSScriptEngine;
+
+/**
+ * GraalJS Script Engine implementation
+ *
+ * @author Jonathan Gilbert - Initial contribution
+ */
+public class OpenhabGraalJSScriptEngine extends InvocationInterceptingScriptEngineWithInvocable<GraalJSScriptEngine> {
+
+    private static final Logger logger = LoggerFactory.getLogger(OpenhabGraalJSScriptEngine.class);
+
+    private static final String REQUIRE_WRAPPER_NAME = "__wraprequire__";
+    private static final String MODULE_DIR = String.join(File.separator, OpenHAB.getConfigFolder(), "automation", "lib",
+            "javascript", "personal");
+
+    // these fields start as null because they are populated on first use
+    @NonNullByDefault({})
+    private String engineIdentifier;
+    @NonNullByDefault({})
+    private Consumer<String> scriptDependencyListener;
+
+    private boolean initialized = false;
+
+    /**
+     * Creates an implementation of ScriptEngine (& Invocable), wrapping the contained engine, that tracks the script
+     * lifecycle and provides hooks for scripts to do so too.
+     */
+    public OpenhabGraalJSScriptEngine() {
+        super(null); // delegate depends on fields not yet initialised, so we cannot set it immediately
+        delegate = GraalJSScriptEngine.create(null,
+                Context.newBuilder("js").allowExperimentalOptions(true).allowAllAccess(true)
+                        .option("js.commonjs-require-cwd", MODULE_DIR).option("js.nashorn-compat", "true") // to ease
+                                                                                                           // migration
+                        .option("js.commonjs-require", "true") // enable CommonJS module support
+                        .fileSystem(new DelegatingFileSystem(FileSystems.getDefault().provider()) {
+                            @Override
+                            public SeekableByteChannel newByteChannel(Path path, Set<? extends OpenOption> options,
+                                    FileAttribute<?>... attrs) throws IOException {
+                                if (scriptDependencyListener != null) {
+                                    scriptDependencyListener.accept(path.toString());
+                                }
+
+                                if (path.toString().endsWith(".js")) {
+                                    return new PrefixedSeekableByteChannel(
+                                            ("require=" + REQUIRE_WRAPPER_NAME + "(require);").getBytes(),
+                                            super.newByteChannel(path, options, attrs));
+                                } else {
+                                    return super.newByteChannel(path, options, attrs);
+                                }
+                            }
+                        }));
+    }
+
+    @Override
+    protected void beforeInvocation() {
+        if (initialized)
+            return;
```

Comment:
Curly braces should always be used with if statements. So please again check your IDE settings.

```suggestion
        if (initialized) {
            return;
        }
```